Direct transpose and overloaded `LinearAlgebra.mul!`

ArrogantGao opened the pull request:
In this PR, the following changes have been made.
1. I added direct implementation for `CuMatrix` type and `CTranspose` type, by treating them as col-major and row-major matrices in `C++`, here are some simple benchmark results:
```
FP32_maxplus_NN: 12.462717 TFLOPS
time per iteration: 11.028009 ms
 
FP32_maxplus_NT: 12.800955 TFLOPS
time per iteration: 10.736617 ms
 
FP32_maxplus_TN: 11.372505 TFLOPS
time per iteration: 12.085196 ms
 
FP32_maxplus_TT: 12.784788 TFLOPS
time per iteration: 10.750195 ms
 
FP32_plusmul_NN: 14.605625 TFLOPS
time per iteration: 9.410002 ms
 
FP32_plusmul_NT: 14.994160 TFLOPS
time per iteration: 9.166165 ms

FP32_plusmul_TN: 12.833895 TFLOPS
time per iteration: 10.709060 ms
 
FP32_plusmul_TT: 14.752284 TFLOPS
time per iteration: 9.316453 ms
 
FP64_plusmul_NN: 6.325359 TFLOPS
time per iteration: 21.728247 ms
 
FP64_plusmul_NT: 6.494176 TFLOPS
time per iteration: 21.163416 ms
 
FP64_plusmul_TN: 4.888482 TFLOPS
time per iteration: 28.114857 ms
 
FP64_plusmul_TT: 6.264414 TFLOPS
time per iteration: 21.939636 ms
```
which shows that for four cases, FP32_maxplus's performance is about 12~13TFLOPs.
In this way, there is no need of any permutations or copy.

2. The whole project is rebuilt using the `TropicalNumbers.jl` package and the `@eval` macro, which made the life much easier. It only took less *44* lines to build up everything. All tests passed on the server.
```
     Testing Running tests...
Test Summary:     | Pass  Total   Time
CuTropicalGEMM.jl |  264    264  28.4s
     Testing CuTropicalGEMM tests passed 
```
PS: I submit a new PR to `TropicalNumbes.jl`, which added some `Int` types needed here.

3. I overloaded `LinearAlgebra.mul!`, which is also tested.

4. The `TropicalMinPlus` algebra is not added here, there are some strange bug I did not figure out why. The only difference between the `C++` code of `TropicalMinPlus` and `TropicalMaxPlus` are the macro, but it keep returning back tons of zeros. I suspect there maybe something wrong with the `min` in GPU. Since `MinPlus` and `MaxPlus` can be replaced by each other easily, I decided to remove it.

GiggleLiu commented on test/test_gemms.jl:
In Julia community, we often use the same file name as the corresponding file in src. Filenames starting with `test_` is more likely a python convension, where pytests will detect test files based on names.

GiggleLiu commented on test/test_gemms.jl:
```diff
@@ -0,0 +1,59 @@
+@testset "Testing the gemms" begin
+    for (MT, DT) in [(Real, [Float32, Float64, Int32, Int64]), (TropicalAndOr, [Bool]), (TropicalMaxPlus, [Float32, Float64]), (TropicalMaxMul, [Float32, Float64])]
+        for T in DT
+            for (M, N, K) in [(0, 0, 0), (2, 0, 0), (2, 2, 0), (5, 6, 7), (101, 102, 103), (213, 145, 274)]
```
Tests look more compact now, good move!

GiggleLiu commented on src/tropical_gemms.jl:
```diff
@@ -0,0 +1,44 @@
+    return MA, NB, KA
+end
+
+for (TA, CTA, tA) in [(:CuMatrix, :CuMatrix, 'N'), (:CTranspose, :(Transpose{<:Any, <:StridedCuVecOrMat}), 'T')]
```
CTA and CTB are not used.

GiggleLiu replied:
When I try using the new GEMM in GenericTensorNetworks, I find the following error:
```julia
cuda patch: Error During Test at /home/leo/.julia/dev/GenericTensorNetworks/test/cuda.jl:8
  Got exception outside of a @test
  MethodError: no method matching matmul!(::LinearAlgebra.Transpose{CountingTropicalF64, CuArray{CountingTropicalF64, 2, CUDA.Mem.DeviceBuffer}}, ::LinearAlgebra.Transpose{CountingTropicalF64, CuArray{CountingTropicalF64, 2, CUDA.Mem.DeviceBuffer}}, ::CuArray{CountingTropicalF64, 2, CUDA.Mem.DeviceBuffer})
```

This is probably due to the incorrect dispatch.

GiggleLiu reviewed (commented):
Good job!

GiggleLiu commented:
Please merge the main branch to make the file change more clear, which should make the review easier.

GiggleLiu commented:
> The TropicalMinPlus algebra is not added here, there are some strange bug I did not figure out why. The only difference between the C++ code of TropicalMinPlus and TropicalMaxPlus are the macro, but it keep returning back tons of zeros. I suspect there maybe something wrong with the min in GPU. Since MinPlus and MaxPlus can be replaced by each other easily, I decided to remove it.

Instead of removing it, what we can do is:
1. Write an issue to describe the problem.
2. Manually throw an proper error message instead of completely removing it, someone might be able look into it and fix the problem.

GiggleLiu commented on src/tropical_gemms.jl:
```diff
@@ -0,0 +1,44 @@
+    for TB in [:CuMatrix, :CTranspose]
+        @eval function LinearAlgebra.mul!(C::CuMatrix{T}, A::$TA{T}, B::$TB{T}) where {T}
+            matmul!(A, B, C)
+            return nothing
```
The `mul!` interface should contain two numbers alpha and beta, such as in
https://github.com/TensorBFS/TropicalGEMM.jl/blob/846b3541b4c98025e79977aa29a04d45715d7f77/src/gemm.jl#L166C44-L166C44

This is very important.

ArrogantGao commented:
The problems mentioned above have been fixed.

The `C` lib can not work properly with 1d vectors, so here I removed the `CTA` and `CTB`.

Tests passed for the revised code.

ArrogantGao commented:
> Please merge the main branch to make the file change more clear, which should make the review easier.

Sorry about that, I forgot to pull before creating this branch.

ArrogantGao commented:
> I just registered a patch version 0.6.1 for `TropicalNumbers`. If you want to force this repo to depend on the correct version, please change the corresponding item in the[compat] section of Project.toml to
> 
> ```
> TropicalNumbers = "0.6.1"
> ```
> 
> [JuliaRegistries/General#92177](https://github.com/JuliaRegistries/General/pull/92177)

Thanks a lot. I will add that to the deps after it is available.

GiggleLiu commented:
> The problems mentioned above have been fixed.
> 
> The `C` lib can not work properly with 1d vectors, so here I removed the `CTA` and `CTB`.

There is not difference between matrix and vector to C. In julia, the vector can be treated as Matrix
```julia
julia> v = randn(4)
4-element Vector{Float64}:
 -0.09222904137882743
 -0.5984111181450423
 -0.49215429693304924
 -0.11885143347454741

julia> size(v, 1)
4

julia> size(v, 2)
1
```

So
* `CuMatrix` -> `CuVecOrMat`
* `const CTranspose{T} = Transpose{T, <:CuVecOrMat{T}}`

GiggleLiu commented on src/tropical_gemms.jl:
```diff
@@ -0,0 +1,44 @@
+# overload the LinearAlgebra.mul!
+for TA in [:CuMatrix, :CTranspose]
+    for TB in [:CuMatrix, :CTranspose]
+        @eval function LinearAlgebra.mul!(C::CuMatrix{T}, A::$TA{T}, B::$TB{T}) where {T}
```
Please add two factors alpha and beta. This interface is for
```julia
C = alpha * A * B + beta * C
```

We should overload this lower level interface.

GiggleLiu replied:
Oh, you did not specify the `T`. Then it applies on CuArrays with any element type. Please specify the element types here to `Union{THE SUPPORTED ELEMENT TYPE LIST}`

ArrogantGao replied:
Oh, I will fix this error.

GiggleLiu commented:
This is the tests I am running, please add it to your testsets
```julia
@testset "cuda patch" begin
    for T in [Tropical{Float64}, CountingTropical{Float64,Float64}]
        a = T.(CUDA.randn(4, 4))
        b = T.(CUDA.randn(4))
        for A in [transpose(a), a, transpose(b)]
            for B in [transpose(a), a, b]
                if !(size(A) == (1,4) && size(B) == (4,))
                    res0 = Array(A) * Array(B)
                    res1 = A * B
                    res2 = mul!(CUDA.zeros(T, size(res0)...), A, B, true, false)
                    @test Array(res1) ≈ res0
                    @test Array(res2) ≈ res0
                end
            end
        end
    end
end
```

GiggleLiu commented:
The tropical gemm seems to be quite unstable:
```julia
  Expression: Array(res1) ≈ res0
   Evaluated: Tropical{Float64}[0.0ₜ 0.30073046684265137ₜ 1.5566071160137653ₜ 0.7655232548713684ₜ; 0.30073046684265137ₜ 1.385499358177185ₜ 1.3180376291275024ₜ 1.850292146205902ₜ; 1.5566071160137653ₜ 1.3180376291275024ₜ 3.1759586334228516ₜ 1.7828304171562195ₜ; 0.7655232548713684ₜ 1.850292146205902ₜ 1.7828304171562195ₜ 2.315084934234619ₜ] ≈ Tropical{Float64}[-0.06274440139532089ₜ 0.30073046684265137ₜ 1.5566071160137653ₜ 0.7655232548713684ₜ; 0.30073046684265137ₜ 1.385499358177185ₜ 1.3180376291275024ₜ 1.850292146205902ₜ; 1.5566071160137653ₜ 1.3180376291275024ₜ 3.1759586334228516ₜ 1.7828304171562195ₜ; 0.7655232548713684ₜ 1.850292146205902ₜ 1.7828304171562195ₜ 2.315084934234619ₜ]
```

ArrogantGao commented:
The following changes have been made.

1. `alpha` and `beta` are added as parameters as the function by changing the `.cu` file as
```
C[OFFSET_col(row, col, M)] = OPERATOR_ADD(OPERATOR_MUL(C[OFFSET_col(row, col, M)], beta), OPERATOR_MUL(accum[OFFSET_row(thread_y, thread_x, THREAD_SIZE_N)], alpha));
```
and now all the different algebras are written in only one file `tropicalgemm_kernels.cu` and compiled into different libs with different macros. This will make the future enhancement easier.

Related tests are added
```
para_set = !(MT == TropicalAndOr) ? [(one(TT), one(TT)), (zero(TT), zero(TT)), (TT(2), TT(3))] : [(one(TT), one(TT)), (zero(TT), zero(TT)), (TT(true), TT(false))]
for (α, β) in para_set
    if MT == TropicalMaxMul
        A = TA == 'T' ? transpose(MT.(CuArray( T.((abs.(rand(T, K, M))).% 1000) ))) : MT.(CuArray( T.((abs.(rand(T, M, K))).% (1000))))
        B = TB == 'T' ? transpose(MT.(CuArray( T.((abs.(rand(T, N, K))).% 1000) ))) : MT.(CuArray( T.((abs.(rand(T, K, N))).% (1000))))
        C = MT.(CuArray( T.((abs.(rand(T, M, N))).%1000)))
    else
        A = TA == 'T' ? transpose(MT.(CuArray(rand(T, K, M)))) : MT.(CuArray(rand(T, M, K)))
        B = TB == 'T' ? transpose(MT.(CuArray(rand(T, N, K)))) : MT.(CuArray(rand(T, K, N)))
        C = MT.(CuArray(rand(T, M, N)))
    end


    hA = Array(A)
    hB = Array(B)
    hC = Array(C)

    C = LinearAlgebra.mul!(C, A, B, α, β)

    hC .= α .* hA * hB .+ β .* hC

    @test Array(C) ≈ hC
end
```

I also made some changes for the `cuda patch` part, the original one is not working corretly
@testset "cuda patch" begin
    for (MT, DT) in [(Real, [Float32, Float64, Int32, Int64]), (TropicalAndOr, [Bool]), (TropicalMaxPlus, [Float32, Float64]), (TropicalMaxMul, [Float32, Float64])]
        for T in DT
            a = MT.(CUDA.rand(T, 4, 4))
            b = MT.(CUDA.rand(T, 4))
            for A in [transpose(a), a, transpose(b)]
                for B in [transpose(a), a, b]
                    testname = "Type " * string(T) * ", size: " * string(size(A)) * " " * string(size(B))
                    @testset "$testname" begin
                        if !(size(A) == (1,4) && size(B) == (4,))
                            res0 = Array(A) * Array(B)
                            res1 = A * B
                            res2 = LinearAlgebra.mul!(MT.(CUDA.zeros(T, size(res0)...)), A, B)
                            @test Array(res1) ≈ res0
                            @test Array(res2) ≈ res0
                        end
                    end
                end
            end
        end
    end
end
```

All tests passed
```
     Testing Running tests...
Test Summary:     | Pass  Total     Time
CuTropicalGEMM.jl |  804    804  1m00.8s
     Testing CuTropicalGEMM tests passed 
```

ArrogantGao commented:
2. The input type of the function have been changed into `CuVecOrMat`.

ArrogantGao commented:
3. I overloaded both `mut!(C, A, B, alpha, beta)` and `mut!(C, A, B)` via
```
for TA in [:CuVecOrMat, :CTranspose]
    for TB in [:CuVecOrMat, :CTranspose]
        @eval function LinearAlgebra.mul!(C::CuMatrix{T}, A::$TA{T}, B::$TB{T}, α::T, β::T) where {T <: Union{Float32, Float64, Int32, Int64, TropicalAndOr, TropicalMaxPlusF32, TropicalMaxPlusF64, TropicalMaxMulF32, TropicalMaxMulF64, TropicalMaxMulI32, TropicalMaxMulI64}}
            C = matmul!(A, B, C, α, β)
            return C
        end
        @eval function LinearAlgebra.mul!(C::CuMatrix{T}, A::$TA{T}, B::$TB{T}) where {T <: Union{Float32, Float64, Int32, Int64, TropicalAndOr, TropicalMaxPlusF32, TropicalMaxPlusF64, TropicalMaxMulF32, TropicalMaxMulF64, TropicalMaxMulI32, TropicalMaxMulI64}}
            C = matmul!(A, B, C, one(T), zero(T))
            return C
        end
    end
end
```

ArrogantGao commented:
> > The TropicalMinPlus algebra is not added here, there are some strange bug I did not figure out why. The only difference between the C++ code of TropicalMinPlus and TropicalMaxPlus are the macro, but it keep returning back tons of zeros. I suspect there maybe something wrong with the min in GPU. Since MinPlus and MaxPlus can be replaced by each other easily, I decided to remove it.
> 
> This is super wield. Instead of removing it, what we can do is:
> 
> 1. Write an issue to describe the problem.
> 2. Manually throw an proper error message instead of completely removing it, someone might be able look into it and fix the problem.

Got it.

ArrogantGao commented:
I fixed the error by `rmul!`, which can not take `0d` matrices as input.
The bug is indeed similar to what happened in `MinPlus` algebra, I will have a check of it.

ArrogantGao commented:
The bug is fixed.
In the previous version, I used `TYPE accum[THREAD_SIZE_M * THREAD_SIZE_N] = {PADDING}` to init the registers, but it turns out that this line will only set the first value of the accum as `PADDING` and all the other value as the default `0`.
In this version I am using a loop to make sure all values in `accum` are correctly init as `PADDING`.

I also added `TropicalMinPlus` into the package, and changed the interface as `CuTropicalGEMM.matmul!(C, A, B, alpha, beta)`.

Here are the revised testsets:
```
@testset "cuda patch positive and negative" begin
    for (MT, DT) in [(TropicalMaxPlus, [Float32, Float64]), (TropicalMinPlus, [Float32, Float64])]
        for T in DT
            for a in [MT.(CUDA.randn(T, 4, 4)), MT.( - CUDA.randn(T, 4, 4))]
                for b in [MT.(CUDA.randn(T, 4)), MT.( - CUDA.randn(T, 4))]
                    for A in [transpose(a), a, transpose(b)]
                        for B in [transpose(a), a, b]
                            testname = "Type " * string(T) * ", size: " * string(size(A)) * " " * string(size(B))
                            @testset "$testname" begin
                                if !(size(A) == (1,4) && size(B) == (4,))
                                    res0 = Array(A) * Array(B)
                                    res1 = A * B
                                    res2 = LinearAlgebra.mul!(MT.(CUDA.zeros(T, size(res0)...)), A, B)
                                    @test Array(res1) ≈ res0
                                    @test Array(res2) ≈ res0
                                end
                            end
                        end
                    end
                end
            end
        end
    end
end
```

All tests passed.
```
     Testing Running tests...
Test Summary:     | Pass  Total     Time
CuTropicalGEMM.jl | 1084   1084  1m01.0s
     Testing CuTropicalGEMM tests passed 
```

ArrogantGao commented:
I updated the `deps/build.jl` and now the package is using `cmake` to compile the deps. 
I tried to directly add the package using `Pkg` from github on another server and tested it, all tests passed.
```
(base) ┌─[xzgao@ganzecheng-amat] - [~/temp] - [Wed Sep 27, 14:55]
└─[$] <> julia test_CuTropicalGemm.jl
Test Summary:     | Pass  Total   Time
Testing the gemms |  780    780  25.5s
Test Summary:                    | Pass  Total   Time
cuda patch positive and negative |  256    256  14.7s
Test Summary:            | Pass  Total  Time
cuda patch positive only |   48     48  3.7s
```

ArrogantGao commented:
Added readme and benchmark file in `Julia`, remove the CUDA benchmark files.

GiggleLiu commented:
I notice that the current version requires cmake rather than nvcc. It errors if I do not have cmake installed.

ArrogantGao commented:
> I notice that the current build requires cmake rather than nvcc. It errors if I do not have cmake installed.

I changed the compiling tool as `cmake` because I found most of the package using `BinaryBuilder.jl` to build deps are written in `cmake`, and currently I am still trying to compile the deps using `BinaryBuilder.jl`.

GiggleLiu commented:
I see. I notice that although the code is much faster, the current version still can not produce the correct result in the following test case.
The output is probabilistic, hence it is very likely you did not sync threads after some computation. Can you please help me to make the following code produce correct result?

```julia
using GenericTensorNetworks, GenericTensorNetworks.Graphs
using CUDA
g = Graphs.random_regular_graph(200, 3)
item(x::AbstractArray) = Array(x)[]
optimizer = TreeSA(ntrials=3)
gp = IndependentSet(g; optimizer=optimizer)
contraction_complexity(gp)
@time CUDA.@sync solve(gp, SizeMax(); usecuda=true, T=Float32)
using CuTropicalGEMM
@time CUDA.@sync solve(gp, SizeMax(); usecuda=true, T=Float32)
```

ArrogantGao commented:
> I see. I notice that although the code is much faster, the current version still can not produce the correct result in the following test case. The output is probabilistic, hence it is very likely you did not sync threads after some computation. Can you please help me to make the following code produce correct result?
> 
> ```julia
> using GenericTensorNetworks, GenericTensorNetworks.Graphs
> using CUDA
> g = Graphs.random_regular_graph(200, 3)
> optimizer = TreeSA(ntrials=3)
> gp = IndependentSet(g; optimizer=optimizer)
> contraction_complexity(gp)
> @time CUDA.@sync solve(gp, SizeMax(); usecuda=true, T=Float32)
> using CuTropicalGEMM
> # If you run the following line multiple times, the result changes.
> @time CUDA.@sync solve(gp, SizeMax(); usecuda=true, T=Float32)
> ```

Yes I also found that quite strange, it seems that the `CUDA.@sync` command is not working properly.
I found that @benchmark somehow can give a correct result when testing simple `matmul` operations as I showed in the README file.
```julia
julia> using TropicalNumbers, CUDA, BenchmarkTools, LinearAlgebra

julia> a = Tropical.(CUDA.randn(4096, 4096));

julia> @benchmark CUDA.@sync $a * $a
BenchmarkTools.Trial: 44 samples with 1 evaluation.
 Range (min … max):  108.365 ms … 123.031 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     116.051 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   116.289 ms ±   4.390 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%

             ▁▁       ▁▁                     ▁     ▄ ▁  ▁█       
  ▆▁▁▁▁▁▆▆▆▆▆██▁▆▁▁▆▁▆██▁▁▆▁▆▆▆▁▁▆▆▁▁▁▁▆▆▁▁▁▁█▆▆▁▁▁█▆█▆▆██▁▁▁▁▆ ▁
  108 ms           Histogram: frequency by time          123 ms <

 Memory estimate: 5.03 KiB, allocs estimate: 95.

julia> using CuTropicalGEMM

julia> @benchmark CUDA.@sync $a * $a
BenchmarkTools.Trial: 440 samples with 1 evaluation.
 Range (min … max):   8.920 μs … 24.497 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     10.733 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   11.363 ms ± 11.347 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                       ▅▇▂  
  █▅▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄▁▄▄███ ▆
  8.92 μs      Histogram: log(frequency) by time      23.3 ms <

 Memory estimate: 160 bytes, allocs estimate: 4.
```

GiggleLiu commented:
It is very wield that the computed result is incorrect.
The above tensor network program basically consists multiple steps of tensor contraction (calls into GEMM).

ArrogantGao commented:
> It is very wield that the computed result is incorrect.
> The above tensor network program basically consists multiple steps of tensor contraction (calls into GEMM).

The result is the same in my test
```julia
julia> solve(gp, SizeMax(); usecuda=true, T=Float32)
0-dimensional CuArray{Tropical{Float32}, 0, CUDA.Mem.DeviceBuffer}:
89.0ₜ

julia> using CuTropicalGEMM

julia> solve(gp, SizeMax(); usecuda=true, T=Float32)
0-dimensional CuArray{Tropical{Float32}, 0, CUDA.Mem.DeviceBuffer}:
89.0ₜ
```

GiggleLiu commented:
Wield, there must be some quantum fluctuation in the code.
Shall we merge the code first and leave an issue so that the PR can be reviewed easier.

ArrogantGao commented:
Ah, I added a `sync` command` after everything in the `C` code, and it seems correct now:
```
julia> @time CUDA.@sync solve(gp, SizeMax(); usecuda=true, T=Float32)
 34.064155 seconds (61.62 M allocations: 4.008 GiB, 4.44% gc time, 0.13% compilation time)
0-dimensional CuArray{Tropical{Float32}, 0, CUDA.Mem.DeviceBuffer}:
90.0ₜ

julia> using CuTropicalGEMM

julia> @time CUDA.@sync solve(gp, SizeMax(); usecuda=true, T=Float32)
  0.485465 seconds (544.38 k allocations: 33.994 MiB, 6.04% gc time)
0-dimensional CuArray{Tropical{Float32}, 0, CUDA.Mem.DeviceBuffer}:
90.0ₜ

julia> @time CUDA.@sync solve(gp, SizeMax(); usecuda=true, T=Float32)
  0.067329 seconds (134.11 k allocations: 6.533 MiB)
0-dimensional CuArray{Tropical{Float32}, 0, CUDA.Mem.DeviceBuffer}:
90.0ₜ

julia> @time CUDA.@sync solve(gp, SizeMax(); usecuda=true, T=Float32)
  0.053670 seconds (134.11 k allocations: 6.533 MiB)
0-dimensional CuArray{Tropical{Float32}, 0, CUDA.Mem.DeviceBuffer}:
90.0ₜ
```
and the `@time` macro is working correctly.
```
julia> @time CUDA.@sync A * B

julia> @time CUDA.@sync A * B;
  0.000115 seconds (6 allocations: 240 bytes)

julia> @time CUDA.@sync A * B;
  0.000132 seconds (6 allocations: 240 bytes)
```

GiggleLiu commented:
I already merged the PR so that we could work on a cleaner branch. Please remember to switch the working branch.
If you have already made the change, please use `git stash` or `git checkout` to save your work.
